File: p51lee/w04/w04_pgs_bs_immigration.scala
object Solution {
    def solution(n: Int, times: Vector[Int]): Long = suppl(times, n, 1L, times.max.toLong * n.toLong, 0)

    def suppl(times: Vector[Int], n: Int, left: Long, right: Long, ans: Long): Long = (left > right) match {
        case true => ans
        case _ => {
            val mid = (left + right) / 2
            if (times.foldLeft(0L) { (acc, time) => acc + mid / time.toLong } < n.toLong)
                suppl(times, n, mid+1, right, ans)
            else
                suppl(times, n, left, mid-1, mid)
        }
    }
}

File: p51lee/w04/w04_pgs_bs_immigration.py
def greedy_solution(n, times):
    times_tot = times[:]
    for _ in range(n):
        min_index = times_tot.index(min(times_tot))
        times_tot[min_index] += times[min_index]
    for i, time in enumerate(times):
        times_tot[i] -= time
    return max(times_tot)
